src/Razor/src/Microsoft.NET.Sdk.Razor/build/netstandard2.0/Microsoft.NET.Sdk.Razor.MvcApplicationPartsDiscovery.targets

```diff
@@ -0,0 +1,103 @@
+<!--
+***********************************************************************************************
+Microsoft.NET.Sdk.Razor.ApplicationPartsDiscovery
+
+WARNING:  DO NOT MODIFY this file unless you are knowledgeable about MSBuild and have
+          created a backup copy.  Incorrect changes to this file will make it
+          impossible to load or build your projects from the command-line or the IDE.
+
+Copyright (c) .NET Foundation. All rights reserved.
+***********************************************************************************************
+-->
+
+<Project ToolsVersion="14.0">
+  <UsingTask
+    TaskName="Microsoft.AspNetCore.Razor.Tasks.DiscoverMvcApplicationParts"
+    AssemblyFile="$(RazorSdkBuildTasksAssembly)"
+    Condition="'$(RazorSdkBuildTasksAssembly)' != ''" />
+
+  <PropertyGroup>
+    <GenerateMvcApplicationPartsAttribute Condition="'$(GenerateMvcApplicationPartsAssemblyAttributes)' == '' AND '$(OutputType)' == 'Exe'">true</GenerateMvcApplicationPartsAttribute>
+
+    <CoreCompileDependsOn Condition="'$(GenerateMvcApplicationPartsAttribute)' == 'true' AND '$(DesignTimeBuild)' != 'true'">
+      _DiscoverMvcApplicationParts;
+      $(CoreCompileDependsOn);
+    </CoreCompileDependsOn>
+
+    <_MvcApplicationPartAttributeGeneratedFile>$(IntermediateOutputPath)$(TargetName).MvcApplicationPartsAssemblyInfo$(DefaultLanguageSourceExtension)</_MvcApplicationPartAttributeGeneratedFile>
+
+    <_DiscoverMvcApplicationPartsDependsOn>
+      ResolveAssemblyReferences;
+      _InitializeDiscoverMvcApplicationParts;
+    </_DiscoverMvcApplicationPartsDependsOn>
+  </PropertyGroup>
+
+  <Target Name="_InitializeDiscoverMvcApplicationParts">
+    <!-- Can be used by testassets which need to simulate a reference to Microsoft.AspNetCore.App -->
+    <PropertyGroup Condition="'$(_ReferencesMicrosoftAspNetCoreApp)' == ''">
+      <_ReferencesMicrosoftAspNetCoreApp>@(FrameworkReference->AnyHasMetadataValue('Identity', 'Microsoft.AspNetCore.App'))</_ReferencesMicrosoftAspNetCoreApp>
+    </PropertyGroup>
+  </Target>
+
+  <Target
+    Name="_DiscoverMvcApplicationParts"
+    Inputs="$(MSBuildAllProjects);$(ResolveAssemblyReferencesStateFile)"
```

Comment:
I saw this being used elsewhere to vary-by RAR's results: https://github.com/dotnet/sdk/blob/506c69fa9f9501ba5a311886256a69c794f17f80/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.GenerateSupportedRuntime.targets#L44. Presumably this is the right property to use.

Reply:
That's actually not considered a good pattern. We have a bug to remove that. There seem to be cases where the state file changes unexpectedly and it can also be disabled via DisableRarCache=true.

Can the Inputs be @(ReferencePath). I guess that's a lot more timestamps to check, but it is more obviously correct.

@rainersigwald Thoughts here?

Reply:
Agreed, this isn't a good approach. The state file is an internal implementation detail of RAR and we can change it without changing the output of RAR, and vice versa.

`@(ReferencePath)` is probably the most-correct initial approach. We bail out when finding the first timestamp that's newer so it's not _always_ a ton more timestamp checking.

Where this can go wrong is when something is _removed_ from `@(ReferencePath)`, but `$(MSBuildAllProjects)` should handle that for most cases, and if not there are possible fixes (hash the list, like 
https://github.com/microsoft/msbuild/blob/0f3d8e94ac11dd892c0c2fadb28b1a42e0d45fb1/src/Tasks/Microsoft.Common.CurrentVersion.targets#L3416-L3435

(if you wind up needing that, maybe we should split that big hash into compile items, references, and "etc" so you can use just the references)

Reply:
I can switch to `@(ReferencePath)`. Would adding `$(ProjectAssetsFile)` to the mix help avoid some of the cases where the project file remains unchanged but references change (for instance when your package references have wildcard)?

Reply:
Essentially:

`Inputs="$(MSBuildAllProjects);$(ProjectAssetsFile);@(ReferencePath)"`

Csc does not hash `@(ReferencePath)` today, so we wouldn't be the first ones to run in to any lingering bugs which would require the itemgroup to be hashed :)

Reply:
That's a good idea. My only hesitation is that I don't think anything else does it today, which might mean there's some hidden problem.

Csc does hash ReferencePath, that's the target I linked above. We just do it in common.targets instead of specifically for C#.

Reply:
We have $(MSBuildAllProjects);$(ProjectAssetsFile) as the heuristic for deps file generation incrementality heuristic. One of those is usually changed when a reference is added or removed. There are ways to defeat it, but it's been considered good enough. So I think there is some prior art for that.

Reply:
> Csc does hash ReferencePath, that's the target I linked above. We just do it in common.targets instead of specifically for C#.

Ahh, oops. Completely missed that. Fair enough, I'll stick to what we know works and hash the inputs.

Reply:
https://github.com/dotnet/sdk/blob/156d53187c9104396673e92e8008ce354f221d8e/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets#L133

Reply:
Ah, awesome, then I have no reservations about that approach. As Nick said there are ways it's not good enough but it's Mostly Fine™ and also Not Worse Than Other Things™.